### ISSUE TITLE
Failed to uplaod .appx packaged file to Microsoft Store

### ISSUE THREAD

**Issue by milkcoke** [fyne-io/fyne#3742]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

## appxmanifest.xml
I made .appx package using `fyne release` CLI

This is example generated appxmanifest.xml on packaging processing 

```xml
<?xml version="1.0" encoding="utf-8"?>
<Package xmlns="http://schemas.microsoft.com/appx/2010/manifest">
    <Identity Name="{{.AppID}}"
              Version="{{.Version}}"
              Publisher="{{.Developer}}" />
    <Properties>
        <DisplayName>{{.Name}}</DisplayName>
        <PublisherDisplayName>{{.DeveloperName}}</PublisherDisplayName>  <!-- ⚠️ 1. Error occurs here -->
        <Logo>Icon.png</Logo>
    </Properties>
    <Prerequisites>
        <OSMinVersion>6.0</OSMinVersion> <!-- ⚠️ 2. Error occurs here -->
        <OSMaxVersionTested>10.0</OSMaxVersionTested>
    </Prerequisites>
    <Resources>
        <Resource Language="en-us" />
    </Resources>

    <!-- <Applications>  ⚠️ 3. Error occurs-->
</Package>
```

### 1. PublisherDisplayName is not correct
In appxmanifest.xml is generated on release processing and which is deleted after .appx file is generated.
However, `PublisherDisplayName` is determined on [this code](https://github.com/fyne-io/fyne/blob/19996646fc4743fcfd06ee2653942e34b6cbddb4/cmd/fyne/internal/commands/release.go#LL350C25-L350C25)
This format is forced to "CN={value}" but we just only input CLI the `-developer  CN=${value}`, value is not the same of PublisherName in taht Microsoft store defined.

### release.go
```go
func (r *Releaser) nameFromCertInfo(info string) string {
	// format should be "CN=Company, O=Company, L=City, S=State, C=Country"
	parts := strings.Split(info, ",")
	cn := parts[0]
	pos := strings.Index(strings.ToUpper(cn), "CN=") // Just parse CN={value} but value is not the same in microsoft store defined. So error occurs.
	if pos == -1 {
		return cn // not what we were expecting, but should be OK
	}

	return cn[pos+3:]
}
```

### 2. OS Minversion is always 6.0
This is not changed on CLI, hardcoded as 6.0

### 3. <Application> tag is not defined
How to modify the tag information? that tag is required.

### CLI command
I try to packaging with the command 

```bash
$ fyne release --target windows \
--appVersion 1.0.0 --appBuild 1 \
--appID [My-App-ID] \
--name [My-App-Name]
--icon [My-App-Icon-Path]
--developer "CN=[MyProductPublisherID]"  // ⚠️ Error occurs here too
--certificate [My-pfx-file]
-password [My-Password]
```

From this, generated .xml file is not appropriate to register on Microsoft Store as app package file

## Product identity
Name, Publisher, PublisherDisplayName are already defiend , this is guided on Microsoft Partner Center > Apps and games > MyProduct
![image](https://user-images.githubusercontent.com/48945177/225907730-c9914a27-dd12-459a-8e32-539e0534bcc9.png)

## Screenshots to upload .appx file
![package_error_log](https://user-images.githubusercontent.com/48945177/225906503-fd3c4bb5-d1a2-4d55-9aad-04f976e8ee5d.png)

### Example code

Refer to [this code](https://github.com/fyne-io/fyne/blob/19996646fc4743fcfd06ee2653942e34b6cbddb4/cmd/fyne/internal/commands/release.go#LL346-L354C69)

There's no processing about `<Application>` tag 
And it seems that microsoft store .appx, .msix package just allow program entrypoint as C#, VisualBasic..

```go
// ..
manifestData := struct{ AppID, Developer, DeveloperName, Name, Version string }{
  AppID: r.AppID,
  Developer:     r.developer,
  DeveloperName: r.nameFromCertInfo(r.developer), // ⚠️ here is problem
  Name:          r.Name,
  Version:       r.combinedVersion(),
}
err = templates.AppxManifestWindows.Execute(manifest, manifestData)
manifest.Close()
if err != nil {
  return errors.New("failed to write application manifest template")
}

// ..
util.CopyFile(r.icon, filepath.Join(payload, "Icon.png"))
util.CopyFile(r.Name, filepath.Join(payload, r.Name))
```

### Fyne version

2.3.1

### Go compiler version

1.20.1

### Operating system and version

Windows 10

### Additional Information

I tried many ways for publish my app on Microsoft store.
However **It seems that just C# , Visua basic program can be registerd succesfully on there.**

This is the sample of AppxManifest.xml provided by Visual Studio.
```xml
<?xml version="1.0" encoding="utf-8"?>

<Package
  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
  xmlns:rescap="http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities"
  IgnorableNamespaces="uap rescap">

  <Identity
    Name="[My App Name]"
    Publisher="CN=[My Publisher ID]"
    Version="1.0.0.0" />

  <Properties>
    <DisplayName>My App</DisplayName>
    <PublisherDisplayName>My Name</PublisherDisplayName>
	<Logo>Images\app_icon.png</Logo>
  </Properties>
	
	<Dependencies>
                <!-- also, MinVersion is just hard coded as 6.0.0.0 -->
		<TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.0.0" MaxVersionTested="10.0.14393.0" />
	</Dependencies>
	
  <Resources>
	<Resource Language="x-generate" />
  </Resources>

  <Applications>
    <Application Id="App"
      Executable="$targetnametoken$.exe"
      EntryPoint="MyApp.App"> <!-- ⚠️ Here is also problem since EntryPoint is just allowed 'Main' method of C# -->
      <uap:VisualElements
        DisplayName="[My App]"
        Description="[My App Description]"
        BackgroundColor="transparent"
        Square150x150Logo="Images\Square150x150Logo.png"
        Square44x44Logo="Images\Square44x44Logo.png">
        <uap:DefaultTile Wide310x150Logo="Images\Wide310x150Logo.png" >
          <uap:ShowNameOnTiles>
            <uap:ShowOn Tile="square150x150Logo"/>
          </uap:ShowNameOnTiles>
        </uap:DefaultTile >
        <uap:SplashScreen Image="Images\SplashScreen.png" />
        <uap:LockScreen BadgeLogo="Images\BadgeLogo.png" Notification="badge"/>
      </uap:VisualElements>
    </Application>
  </Applications>

  <Capabilities>
    <Capability Name="internetClient" />
    <rescap:Capability Name="runFullTrust" />
  </Capabilities>
</Package>
```

I tried input <EntryPoint> main (go program) but it failed.
Visual studio logging error as shown below
![image](https://user-images.githubusercontent.com/48945177/225912681-bc25e070-b409-41d4-b4c9-a17a1b38e6e4.png)

It should be `Main` of C# (I guess)

Please confirm this process. That must be outdated.
**And is there any other solution? for packaging file for uploading microsoft store?**

**Comment by milkcoke** [comment]
@Bluebugs 

### Respond your question
I guess you ask me if would I modify the appxmanifest.xml file and re-package it using [MakeApp.exe](https://learn.microsoft.com/en-us/windows/win32/appxpkg/make-appx-package--makeappx-exe-)? 
**No, I didn't modify appmanifest.xml file from generated .appx by `fyne release`.**
When I unpackage the .appx file , there's a appxmanifest.xml file in `ProgramFiles/WindowsApps` 
the first appxmanifest.xml example is it. 

Microsoft store requires upload .appx or .msix file.

### Describe What I tried
I tried to modify `AppxManifest.xml` manually in my visual studio which refer to fyne application as executable file and entrypoint

last `AppxManifest.xml` example provided by Visual Studio is mine
I stuck on this issue
![image](https://user-images.githubusercontent.com/48945177/226071294-385c7292-a205-4ed3-b790-c439fc2d1868.png)

### My Idea
If there's no way set entrypoint `main` of golang instead of C#, Visual Basic
I'm thinking about making simple C# program including `Main` to execute `MyApplication.exe` file generated by `fyne package`.
It seems not good approach though.

**Comment by milkcoke** [comment]
I'm using Visual Studio 2022 (Community)

File > New > Project from Existing code as shown below
![image](https://user-images.githubusercontent.com/48945177/226089194-02a02424-2f91-418b-9ab7-4f7beaba192d.png)

So I could do import (references) wrapping go fyne project as C# project.
I set `Executable=MyExeFileName.exe` in my application project root.
But `CS5001 : Program does not contain a static 'Main' method suitable for an entry point`

## Package.appxmanifest.xml
![image](https://user-images.githubusercontent.com/48945177/226090520-fb853307-bf2d-42ed-97bb-a373ba022739.png)

I tried to execute the file , It succeed to execute but there's problem.
`MyApp.exe` generated by go + fyne build is not executed. instead, just `static void Main(){}` code is executed not doing anything.
It seems allowing just Executable file is mapped Generated by EntryPoint `Main` of C#.  So I failed to execute my fyne application file.
Please refer to [this docs](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs5001?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(CS5001)%26rd%3Dtrue)

I'm finding any other way in the static Main of C# executing `MyApp.exe` .
Thank your fast response.

**Comment by andydotxyz** [comment]
I'm not sure why this is so broken - it did work in the past when I wrote the packager and the docs for the book.
I wonder if the wrong tag is being used, C apps won't allow a custom entry point and by adding a C# app all you're doing is providing a no-op main function so the app packages, but it won't run the code from the other executable.

I can't figure out what has changed.

**Comment by andydotxyz** [comment]
> It seems allowing just Executable file is mapped Generated by EntryPoint Main of C#. So I failed to execute my fyne application file.
Please refer to this docs

That error / doc refers to the C# compiler, not the packaging format...

**Comment by andydotxyz** [comment]
Yup, the Application manifest requirement of Executable+EntryPoint (or StartPage) is a UWP thing. We need to refer to docs for the C/C++ app manifests.

**Comment by andydotxyz** [comment]
Thanks for looking into this, but you're right we need to find the metadata that will run the app without a runner app included.

**Comment by andydotxyz** [comment]
It seems Microsoft provided a response (fairly basic) - which confirms it should be possible with just the .exe file.

**Comment by milkcoke** [comment]
Hello @andydotxyz ,
Your link is mine.. that was my question you suggest linked.

I succeed publish my app on microsoft store as `.exe` package. However it's not good solution.

I've tried many ways for packaging this app on microsoft store.
I describe what's the problem and What I've tired resolving this problem.

#### Youtube link
https://youtu.be/UOBjJCvV_X4

#### Toolbox Dev beta version
https://apps.microsoft.com/store/detail/XP99Z5LSQPTS2F

I'll wait your response. I need your help.

**Comment by andydotxyz** [comment]
> Your link is mine.. that was my question you suggest linked.

Yes indeed, but as it was not linked to this issue I wanted to mention it for whoever works on this ticket.

**Comment by milkcoke** [comment]
I succeed packaging my golang + fyne.io application as `.msix` and `.appx` both.
And publishing on Microsoft Store.
There's 5 points we need packaging using fyne.io application for windows desktop application.
I didn't try any other device or platform.

## 1. Packaging Format
I recommend `.appx` or `.msix`.
|index|msix|appx|
|------|-----|-----|
|Platform|Multiplatform <br>(xbox, VR device, desktop, windows mobile)| Only for Windows 10 newer|
|Running on|Container|Container|
|Install Folder|%ProgramFiles%\WindowsApps|%ProgramFiles%\WindowsApps|


### Tips for choice
- All current Appx Apps will be 100% supported in MSIX
- msix minimum os version must be > 10.17134.0 (October 2018)
- appx supports all windows 10 version 
#### appxmanifest.xml
```xml
    <Dependencies>
        <!-- if package type is .msix MinVersion must be > 10.17134.0 (October 2018) -->
        <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.17763.0" MaxVersionTested="10.0.22621.0" />
    </Dependencies>
```

### Format conclusion.
If you want to support your app for other devices use `.msix` or just for windows desktop? choose `.appx`


## 2. <Application> tag
Current `fyne release` command doesn't support making `Applications` tag .
This tag is for providing entrypoint of your application after installation is complete.
If you doesn't Set `<Application>` tag on appxmanifest.xml, Even if install completes, app is not searched on your computer because it doesn't know what's the entry point. 
In fact, without application tag, your package should be installed on `%ProgramFiles%\WindowsApps` with `.exe` file build by fyne.io but it can't be accessed without entrypoint although you have administrator 

![image](https://user-images.githubusercontent.com/48945177/232294282-992a517a-4f33-4b8a-a2fc-e0a64ccaf941.png)
Without this tag, Can't be launched since application file is not found on your computer.

### How to figure Application tag on `appxmanifest.xml`?
In current version (fyne 2.3.x), You should configure `appxmanifest.xml`.
It doesn't make `Application` tag for you.

### Example appxmanifest.xml
```xml
<?xml version="1.0" encoding="utf-8"?>
<Package
        xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
        xmlns:rescap="http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities"
        xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
        IgnorableNamespaces="rescap uap">

    <Identity
            Name="<MicrosoftStore Developer  Product ID>"
            Publisher="<Microsoft Developer Publisher ID>"
            Version="<Your version>"
            ProcessorArchitecture="x64"/>

    <Properties>
        <DisplayName>Dev Toolbox</DisplayName>
        <PublisherDisplayName><Microsoft Developer Publisher DisplayName></PublisherDisplayName>
        <Logo>{Your logo file}</Logo>
    </Properties>

    <Dependencies>
        <!-- if package type is .msix MinVersion must be > 10.17134.0 (October 2018) -->
        <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.17763.0" MaxVersionTested="10.0.22621.0" />
    </Dependencies>

    <Resources>
        <Resource Language="en-us"/>
    </Resources>


    <Applications>
        <Application Id="<AppID> (Don't use spacebar)" Executable="<.exe file build on golang + fyne io in package directory>" EntryPoint="Windows.FullTrustApplication">
<!-- VisualElements tag is also required -->
            <uap:VisualElements DisplayName="<Shown on your application on desktop>" Description="<Your description>" BackgroundColor="#000000"
                                Square44x44Logo="images\app_icon_44.png" Square150x150Logo="images\app_icon_150.png">
                <uap:DefaultTile Square71x71Logo="images\app_icon_71.png">
                    <uap:ShowNameOnTiles>
                        <uap:ShowOn Tile="square150x150Logo"/>
                    </uap:ShowNameOnTiles>
                </uap:DefaultTile>
            </uap:VisualElements >
        </Application>
    </Applications>


    <Capabilities>
<!-- This is optional tag, I'm using internet client (download) -->
        <Capability Name="internetClient" />
<!-- This is required tag for supporting win32 app , fyne io is win32 app default.-->
        <rescap:Capability Name="runFullTrust" />
    </Capabilities>
</Package>
```

#### Check on you microsoft > app and games > Product ID

![image](https://user-images.githubusercontent.com/48945177/232295885-cd478487-bc54-41b2-9a75-e10a84f7852d.png)


## 3. Certificate
You should get developer certificate from microsoft .
I recommend get it from Visual Studio.
When you install visual studio it maybe install windows 10 SDK, and 11 SDK with `MakeAppx` and `signtool` for packaging, signing.

#### Package.appxmanifest

![image](https://user-images.githubusercontent.com/48945177/232297764-9a0ba67d-5b54-407e-a88f-68afe9dc8640.png)
1. Connect on your developer account on visual studio. It automatically show self-signed certificate for you.
2. Choose from store
3. Click `View full certificate`
4. Export as file (.pfx) for input command line of fyne
5. Install on `Trusted Root Certificate Authorities`

![image](https://user-images.githubusercontent.com/48945177/232298529-22f766c5-7f34-4f6a-afd4-eb30a82af98e.png)



Also, there's many guide for CLI for getting certificates. refer to [this docs](https://learn.microsoft.com/en-us/dotnet/core/additional-tools/self-signed-certificates-guide)


## 4. Packaging and Sigining

It seems that current `fyne release` command has error for `makeappx` path finding.
I executed manually `makeappx` and `signtool` command.

At first, prepare your package directory like as shown below
![image](https://user-images.githubusercontent.com/48945177/232300115-79a7a663-65e3-47e2-9d3e-008714c738fe.png)

```bash
# you can also packaging with .appx
$ makeappx pack -d [your package directory] /p [package-name].msix
# signing with your certificate (.pfx or .cer)
$ signtool sign /a /v /fd SHA256 /f [certificate file] /p [Certificate Password] [Package file path created above]
```

> **Done! You can upload your `.appx` or `.msix` for submitting on Microsoft Store!**

## Conclusion.
In current, users should configure package directory with `appxmanifest.xml` and icon, exe file build on golang and fyne.io

![image](https://user-images.githubusercontent.com/48945177/232300229-5c82a45d-779f-4615-bc91-f0cd0499238f.png)

I got this helping @andydotxyz . Thanks.

I expect developing fyne release and package process according to required feature of microsoft store.
I'll be the first man of publisher using this fyne.io!!!

**Comment by lordofscripts** [comment]
Still not possible?Just wondering...

**Comment by andydotxyz** [comment]
I'm not sure from what is above made you think it is not possible @lordofscripts.

Did you try `fyne release -os windows` ?

I have not figured out why the OP could not get it to work, but it worked on my setup.
